File: julia/Euler/src/bin/p0055.jl

# project euler: problem 55

module Prob0055

function is_rychrel(n)
    tmp = parse(BigInt, reverse(string(n)))
    for _ = 1:50
        n += tmp
        tmp = parse(BigInt, reverse(string(n)))
        if n == tmp
            return false
        end
    end
    return true
end

function solve_0055(upper::Int = 10_000)
    count(is_rychrel, BigInt(1):(upper -1))
end

end #module

using .Prob0055: solve_0055
export solve_0055

File: clojure/src/project_euler/sol/p0055.clj
(ns project-euler.sol.p0055
  (:require
   [project-euler.lib.math :as math]
   [project-euler.lib.util :as util]))

(defn- rychrel?
  [n]
  (letfn [(reverse-num [n] (-> (util/digits n) (reverse) (util/undigits)))]
    (loop [n n
           tmp (reverse-num n)
           cnt 50]
      (if (zero? cnt)
        true
        (let [next-n (+' n tmp)]
          (if (math/palindrome? next-n)
            false
            (recur next-n (reverse-num next-n) (dec cnt))))))))

(defn solve
  ([]
   (solve 10000))
  ([upper]
   {:pre [(pos? upper)]}
   (->> (range 1 upper)
        (filter rychrel?)
        (count))))
